### ISSUE TITLE
`for_loops_over_fallibles` suggestion wrong in macro context

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust-clippy#16133]
Using the following flags

`--force-warn for_loops_over_fallibles`

this code:

```rust
//@ check-pass
//@ compile-flags: -Afor_loops_over_fallibles -Warray_into_iter

fn main() {
    macro_rules! mac {
        (iter $e:expr) => {
            $e.iter()
        };
        (into_iter $e:expr) => {
            $e.into_iter() //~ WARN this method call resolves to
            //~^ WARN this changes meaning in Rust 2021
        };
        (next $e:expr) => {
            $e.iter().next()
        };
    }

    for _ in dbg!([1, 2]).iter() {}
    for _ in dbg!([1, 2]).into_iter() {} //~ WARN this method call resolves to
    //~^ WARN this changes meaning in Rust 2021
    for _ in mac!(iter [1, 2]) {}
    for _ in mac!(into_iter [1, 2]) {}
    for _ in mac!(next [1, 2]) {}
}

```

caused the following diagnostics:

```
    Checking _macro-expansion-empty-span-147408 v0.1.0 (/tmp/icemaker_global_tempdir.gOhJUmIOFvoe/icemaker_clippyfix_tempdir.rybLW249JEAF/_macro-expansion-empty-span-147408)
warning: for loop over an `Option`. This is more readably written as an `if let` statement
  --> src/main.rs:23:14
   |
23 |     for _ in mac!(next [1, 2]) {}
   |              ^^^^^^^^^^^^^^^^^
   |
   = note: requested on the command line with `--force-warn for-loops-over-fallibles`
help: to iterate over `$e.iter()` remove the call to `next`
   |
14 -             $e.iter().next()
14 +             .by_ref().next()
   |
help: consider using `if let` to clear intent
   |
23 -     for _ in mac!(next [1, 2]) {}
23 +     if let Some(_) = mac!(next [1, 2]) {}
   |

warning: `_macro-expansion-empty-span-147408` (bin "_macro-expansion-empty-span-147408") generated 1 warning
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.33s

```

However after applying these diagnostics, the resulting code:

```rust
//@ check-pass
//@ compile-flags: -Afor_loops_over_fallibles -Warray_into_iter

fn main() {
    macro_rules! mac {
        (iter $e:expr) => {
            $e.iter()
        };
        (into_iter $e:expr) => {
            $e.into_iter() //~ WARN this method call resolves to
            //~^ WARN this changes meaning in Rust 2021
        };
        (next $e:expr) => {
            .by_ref().next()
        };
    }

    for _ in dbg!([1, 2]).iter() {}
    for _ in dbg!([1, 2]).into_iter() {} //~ WARN this method call resolves to
    //~^ WARN this changes meaning in Rust 2021
    for _ in mac!(iter [1, 2]) {}
    for _ in mac!(into_iter [1, 2]) {}
    if let Some(_) = mac!(next [1, 2]) {}
}

```
no longer compiled:

```
    Checking _macro-expansion-empty-span-147408 v0.1.0 (/tmp/icemaker_global_tempdir.gOhJUmIOFvoe/icemaker_clippyfix_tempdir.rybLW249JEAF/_macro-expansion-empty-span-147408)
error: expected expression, found `.`
  --> src/main.rs:14:13
   |
14 |             .by_ref().next()
   |             ^ expected expression
...
23 |     if let Some(_) = mac!(next [1, 2]) {}
   |                      ----------------- in this macro invocation
   |
   = note: this error originates in the macro `mac` (in Nightly builds, run with -Z macro-backtrace for more info)

error: could not compile `_macro-expansion-empty-span-147408` (bin "_macro-expansion-empty-span-147408" test) due to 1 previous error
warning: build failed, waiting for other jobs to finish...
error: could not compile `_macro-expansion-empty-span-147408` (bin "_macro-expansion-empty-span-147408") due to 1 previous error

```

Version:
```
rustc 1.93.0-nightly (4b1b6dde0 2025-11-22)
binary: rustc
commit-hash: 4b1b6dde0c15243857165672a1c616a4523ef32a
commit-date: 2025-11-22
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.5

```

**Comment by profetia** [comment]
`for_loops_over_fallibles` is a compiler lint. You should open this issue in the rust repo instead.

**Comment by matthiaskrgr** [comment]
ah woops ^^
should be dupe of https://github.com/rust-lang/rust/issues/148114 then
